presto-orc/src/main/java/com/facebook/presto/orc/reader/MapFlatStreamReader.java

```diff
@@ -0,0 +1,429 @@
+    private final OrcType.OrcTypeKind keyOrcType;
+
+    private final List<InputStreamSource<BooleanInputStream>> inMapStreamSources = new ArrayList<>();
+    private final List<BooleanInputStream> inMapStreams = new ArrayList<>();
+    private final List<StreamReader> valueStreamReaders = new ArrayList<>();
+    private final List<StreamDescriptor> valueStreamDescriptors = new ArrayList<>();
+
+    private Block dictionaryBlock;
+    private boolean[][] inMapVectors;
+    private int readOffset;
+    private int nextBatchSize;
+
+    @Nonnull
+    private InputStreamSource<BooleanInputStream> presentStreamSource = missingStreamSource(BooleanInputStream.class);
+    @Nullable
+    private BooleanInputStream presentStream;
+
+    private boolean rowGroupOpen;
+
+    public MapFlatStreamReader(StreamDescriptor streamDescriptor, DateTimeZone hiveStorageTimeZone, AggregatedMemoryContext systemMemoryContext)
+    {
+        this.streamDescriptor = requireNonNull(streamDescriptor, "stream is null");
+        this.hiveStorageTimeZone = requireNonNull(hiveStorageTimeZone, "hiveStorageTimeZone is null");
+        this.systemMemoryContext = requireNonNull(systemMemoryContext, "systemMemoryContext is null");
+        this.keyOrcType = streamDescriptor.getNestedStreams().get(0).getStreamType();
+        this.baseValueStreamDescriptor = streamDescriptor.getNestedStreams().get(1);
+    }
+
+    @Override
+    public void prepareNextRead(int batchSize)
+    {
+        readOffset += nextBatchSize;
+        nextBatchSize = batchSize;
+    }
+
+    @Override
+    public Block readBlock(Type type)
+            throws IOException
+    {
+        if (!rowGroupOpen) {
+            openRowGroup();
+        }
+
+        if (readOffset > 0) {
+            if (presentStream != null) {
+                // skip ahead the present bit reader, but count the set bits
+                // and use this as the skip size for the data reader
+                readOffset = presentStream.countBitsSet(readOffset);
+            }
+            if (readOffset > 0) {
+                for (int i = 0; i < valueStreamReaders.size(); i++) {
+                    int valueReadOffset = inMapStreams.get(i).countBitsSet(readOffset);
+                    valueStreamReaders.get(i).prepareNextRead(valueReadOffset);
+                }
+            }
+        }
+
+        boolean[] nullVector = null;
+
+        if (presentStream == null) {
+            for (int i = 0; i < inMapStreams.size(); i++) {
+                if (inMapStreams.get(i) != null) {
+                    inMapVectors[i] = new boolean[nextBatchSize];
+                    inMapStreams.get(i).getSetBits(nextBatchSize, inMapVectors[i]);
+                }
+            }
+        }
+        else {
+            nullVector = new boolean[nextBatchSize];
+            int nullValues = presentStream.getUnsetBits(nextBatchSize, nullVector);
+            if (nullValues != nextBatchSize) {
```

Comment:
what happens if `nullValues == nextBatchSize` (i.e. every map is null)

Reply:
> what happens if `nullValues == nextBatchSize` (i.e. every map is null)

It looks like the code will behave incorrectly in such cases. Improve test coverage?

Reply:
Good call, I'll add that test